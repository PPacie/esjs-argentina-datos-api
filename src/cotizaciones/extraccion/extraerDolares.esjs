import { format } from 'date-fns'

exportar asincrono funcion extraerDolares() {
  const respuesta = esperar fetch('https://dolarapi.com/v1/dolares')

  const dolares = esperar respuesta.json()

  const hoy = crear Fecha()

  retornar dolares.mapear((dolar) => ({
    casa: dolar.casa,
    compra: dolar.compra,
    venta: dolar.venta,
    fecha: format(hoy, 'yyyy-MM-dd'),
    fechaActualizacion: dolar.fechaActualizacion,
  }))
}
