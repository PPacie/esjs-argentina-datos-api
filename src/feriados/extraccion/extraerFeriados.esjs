importar { load } desde 'cheerio'

// Mapear los nombres de los meses a números
const mesesNumeros = {
  Enero: 1,
  Febrero: 2,
  Marzo: 3,
  Abril: 4,
  Mayo: 5,
  Junio: 6,
  Julio: 7,
  Agosto: 8,
  Septiembre: 9,
  Octubre: 10,
  Noviembre: 11,
  Diciembre: 12,
}

exportar asincrono funcion extraerFeriados(año) {
  const url = `https://www.lanacion.com.ar/feriados/${año}`
  const $ = load(esperar (esperar fetch(url)).text())

  // Matriz para almacenar los feriados
  var feriados = []

  // Seleccionar los elementos que contienen los feriados
  $('div.holidays-card-calendar').each((i, div) => {
    // Extraer el mes del feriado
    const mes = $(div).buscar('h3.com-text').text()

    // Iterar sobre cada feriado en el mes
    $(div)
      .buscar('ul.holidays-list li')
      .each((j, li) => {
        // Extraer el día, tipo y nombre del feriado
        const dia = $(li).buscar('span').text()
        const tipo = $(li).buscar('span').attr('class')
        const nombre = $(li).buscar('h4.com-text').text()

        // Añadir el feriado a la matriz
        feriados.agregar({
          dia: Numero(dia),
          mes: mesesNumeros[mes],
          año,
          fecha: crear Fecha(año, mesesNumeros[mes] - 1, Numero(dia))
            .aCadenaISO()
            .dividir('T')[0],
          tipo: interpretarTipo(limpiarTipo(tipo)),
          nombre,
        })
      })
  })

  retornar feriados
}

funcion limpiarTipo(tipo) {
  retornar tipo.reemplazar(/[^a-zA-Z0-9]/g, '') // Eliminar caracteres no alfanuméricos
}

funcion interpretarTipo(tipo) {
  elegir (tipo) {
    caso 'immovable':
      retornar 'inamovible'
    caso 'transferable':
      retornar 'trasladable'
    caso 'bridge':
      retornar 'puente'
    porDefecto:
      retornar 'otro'
  }
}

