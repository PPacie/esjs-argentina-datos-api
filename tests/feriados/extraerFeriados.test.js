import { expect, it, describe } from 'vitest'
import { extraerFeriados } from '@/feriados/extraccion/extraerFeriados'

describe('extraerFeriados', () => {
  it('extrae los feriados del aÃ±o', async () => {
    const feriados = await extraerFeriados(2024)

    expect(feriados).toMatchSnapshot()
  })
})
