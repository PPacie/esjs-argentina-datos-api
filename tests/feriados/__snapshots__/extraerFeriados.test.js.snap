// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extraerFeriados > extrae los feriados del año 1`] = `
[
  {
    "año": 2024,
    "dia": 1,
    "fecha": "2024-01-01",
    "mes": 1,
    "nombre": "Año nuevo",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 12,
    "fecha": "2024-02-12",
    "mes": 2,
    "nombre": "Carnaval",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 13,
    "fecha": "2024-02-13",
    "mes": 2,
    "nombre": "Carnaval",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 24,
    "fecha": "2024-03-24",
    "mes": 3,
    "nombre": "Día Nacional de la Memoria por la Verdad y la Justicia",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 29,
    "fecha": "2024-03-29",
    "mes": 3,
    "nombre": "Viernes Santo",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 1,
    "fecha": "2024-04-01",
    "mes": 4,
    "nombre": "Feriado puente turístico",
    "tipo": "puente",
  },
  {
    "año": 2024,
    "dia": 2,
    "fecha": "2024-04-02",
    "mes": 4,
    "nombre": "Día del Veterano y de los Caídos en la Guerra de Malvinas",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 1,
    "fecha": "2024-05-01",
    "mes": 5,
    "nombre": "Día del Trabajador",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 25,
    "fecha": "2024-05-25",
    "mes": 5,
    "nombre": "Día de la Revolución de Mayo",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 17,
    "fecha": "2024-06-17",
    "mes": 6,
    "nombre": "Paso a la Inmortalidad del General Martín Güemes",
    "tipo": "trasladable",
  },
  {
    "año": 2024,
    "dia": 20,
    "fecha": "2024-06-20",
    "mes": 6,
    "nombre": "Paso a la Inmortalidad del General Manuel Belgrano",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 21,
    "fecha": "2024-06-21",
    "mes": 6,
    "nombre": "Feriado puente turístico",
    "tipo": "puente",
  },
  {
    "año": 2024,
    "dia": 9,
    "fecha": "2024-07-09",
    "mes": 7,
    "nombre": "Día de la Independencia",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 17,
    "fecha": "2024-08-17",
    "mes": 8,
    "nombre": "Paso a la Inmortalidad del Gral. José de San Martín",
    "tipo": "trasladable",
  },
  {
    "año": 2024,
    "dia": 11,
    "fecha": "2024-10-11",
    "mes": 10,
    "nombre": "Feriado puente turístico",
    "tipo": "puente",
  },
  {
    "año": 2024,
    "dia": 12,
    "fecha": "2024-10-12",
    "mes": 10,
    "nombre": "Día del Respeto a la Diversidad Cultural",
    "tipo": "trasladable",
  },
  {
    "año": 2024,
    "dia": 18,
    "fecha": "2024-11-18",
    "mes": 11,
    "nombre": "Día de la Soberanía Nacional",
    "tipo": "trasladable",
  },
  {
    "año": 2024,
    "dia": 8,
    "fecha": "2024-12-08",
    "mes": 12,
    "nombre": "Día de la Inmaculada Concepción de María",
    "tipo": "inamovible",
  },
  {
    "año": 2024,
    "dia": 25,
    "fecha": "2024-12-25",
    "mes": 12,
    "nombre": "Navidad",
    "tipo": "inamovible",
  },
]
`;
